pkg/endpoint/manager.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2016-2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/endpoint/id"
+	"github.com/cilium/cilium/pkg/eventqueue"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+type endpointManager interface {
+	AllocateID(id uint16) (uint16, error)
+	RunK8sCiliumEndpointSync(*Endpoint)
+	UpdateReferences(map[id.PrefixType]string, *Endpoint)
+	UpdateIDReference(*Endpoint)
+	RemoveReferences(map[id.PrefixType]string)
+	RemoveID(uint16)
+	ReleaseID(*Endpoint)
+}
+
+// Expose exposes the endpoint to the endpointmanager. After this function
+// is called, the endpoint may be accessed by any lookup in the endpointmanager.
+func (e *Endpoint) Expose(mgr endpointManager) error {
```

Comment:
My complaint may be out of scope since you were just moving things to more appropriate spots.
I'm not too keen on the way these  Endpoint/EndpointManager functions worked out. I think we now have something -> endpointmanager -> endpoint -> endpointmanager call chains. This is going to be difficult to follow, or at least very unexpected. It's possible that the coupling we have is too deep to undo and that merging endpointmanager into pkg/endpoint is the best choice (tgraf suggested this a while back but I had hoped to use the separation as a way to force clearer boundaries for Endpoint, exactly as you are doing here).
Looking through how we seem to use these functions, it should be possible to do the operation on the endpoint(set the ID, "expose" and "unexpose"), then pass the endpoint itself to the endpoint manager (or in the other order, when doing unexpose). One exception is `EndpointManager.AddEndpoint` where we also do `ep.SetDesiredIngressPolicyEnabled(alwaysEnforce)`. It just so happens that we call that from `Daemon.createEndpoint` and that seems like a better place for those two lines anyway. 

In more concrete terms,
where we would call endpointmanager.Insert we now:
`mgr.AllocateID()`
`ep.SetID()` (or something similar)
`mgr.Insert(ep)` (expose is also good here, I just wanted to make it distinct than the PR implementation)

where we would call endpointmanager.Remove we now:
`mgr.Remove(ep)`
`ep.Delete()` (or whatever we do to turn it off, the order could be the other way too, if we wanted to expose in-delete endpoints).

The lock ordering in  `updateReferences` (in the new code) or functions that wrap it should also be ok. We currently go through each endpoint field with an if statement and update the entries in endpointmanager. This requires holding both locks (ep and epMgr) but those fields on the endpoint are "static" in that they are set or not when we do this. Grabbing them all in one go then doing logic on the values while holding only the endpointmanager lock should be equivalent. Even in the case where we make 4 separate calls, each locking the endpoint, I think it's logically equivalent to us from an operation ordering perspective. The fields are either set together (the containerID, dockerID and containerName) or should be (the IP values). This might be dangerous in the future but we already have this issue today with the current code. We don't race on concurrent field updates but we definitely set the container info separately from the IPs for the endpoint, so we must already call updateReferences more than once somewhere.